File: C(++)/031. Next Permutation.cpp
/*
031. Next Permutation
https://leetcode.com/problems/next-permutation/

cpp 12 ms
*/

/*
Implement next permutation, which rearranges numbers into the lexicographically next greater permutation of numbers.

If such arrangement is not possible, it must rearrange it as the lowest possible order (ie, sorted in ascending order).

The replacement must be in-place, do not allocate extra memory.

Here are some examples. Inputs are in the left-hand column and its corresponding outputs are in the right-hand column.
1,2,3 → 1,3,2
3,2,1 → 1,2,3
1,1,5 → 1,5,1
*/

class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int len=nums.size(),i=len-1;
        if (len==1) return;
        while (i>0&&nums[i-1]>=nums[i])
            i--;
        if (i==0)
        {
            sort(nums.begin(),nums.end());
            return;
        }
        int p=i,mi=nums[i];
        for (int j=i+1;j<len;j++)
            if (mi>nums[j]&&nums[i-1]<nums[j])
                p=j;
        int t=nums[i-1];
        nums[i-1]=nums[p];
        nums[p]=t;
        sort(nums.begin()+i,nums.end());
    }
};

File: Java/031. Next Permutation.java
/*
031. Next Permutation
https://leetcode.com/problems/next-permutation/

java 25 ms
*/

public class Solution {
	public void nextPermutation(int[] nums) {
		int len=nums.length,from=0;
		boolean flag=false;
		// 依次放宽寻找界限，没有外层界限的话例如1,3,4,2会出错
		for (int i=len-2;i>=0&&!flag;i--)
			for (int j=len-1;j>i&&!flag;j--) // 此次考虑哪一位
				for (int k=j-1;k>=i;k--) // 这一位是否比前边某一位大
					if (nums[j]>nums[k]) // 是则交换这两位
					{
						from=k+1;
						int temp=nums[k];
						nums[k]=nums[j];
						nums[j]=temp;
						flag=true;
						break;
					}
		Arrays.sort(nums,from,len); // 重新排序之后的部分
	}
}
